### DISCUSSION TITLE
Arbitrary array combinations that correlate with the type key

### DISCUSSION THREAD

**Discussion by timuradilev** [vimeo/psalm#8230] [Q&A]
In the example below depending on the type's value correct keys are queried.
https://psalm.dev/r/35b374c933
Is Psalm capable of learning that the code is correct? Are there any workarounds for the problem?

**Comment by orklah** [comment]
Currently, Psalm instantly collapse your notation into a single shape: https://psalm.dev/r/5ee0741dc3

I'm not sure what would happen if Psalm stopped doing that, it could be a good experiment if you're up for it. (It's done in TypeCombiner::combine)

**Reply by AndrolGenhald** [reply]
It's not currently possible because `Union` can only hold a single array with the `"array"` type key. This specifically is one of the reasons I started working on rewriting the type system, I'm not a fan of collapsing `array{foo: "bar"}|array{bar: "baz"}` to `array{foo?: "bar", bar?: "baz"}`.

**Reply by orklah** [reply]
Oh right, forgot about that too
